### PR TITLE
Make shape an optional attribute for constant components

### LINKED ISSUE #1663: Constant Records for JoinedArrays: Skip `shape` for Particles
For ADIOS2 JoinedArray support, we need to address this regression:
https://github.com/openPMD/openPMD-standard/pull/289

Opening this issue to document our discussion results two weeks ago @franzpoeschel @guj @ax3l

### PR DESCRIPTION
**Pull request by franzpoeschel**
Fix #1663

TODO:

- [x] Actually test this
- [x] Link to standard update: https://github.com/openPMD/openPMD-standard/pull/289
- [x] Maybe only do this for standard version 2.0 - actually ok as a fallback to expose in 1.X files, too UPDATE: [This comment](https://github.com/openPMD/openPMD-standard/pull/289#issuecomment-2407263974) explains a corner case that makes me hesitate adding this for version 1.0 since it might break datasets in singular cases. UPDATE: Corner case not relevant enough, would imply misuse of reserved `value` attribute, so not strictly standard-breaking.
- [x] For particles only or for meshes too?
- [x] Test that the shape can be queried from every component by the reader, even if shape attribute is missing
- [x] Document compatibility option

### CODE REVIEW

**Comment by guj** [comment]
Thanks I will give it a try!

**Review by guj — APPROVED** [review]
I just tried with joined array, it worked!

File: .clang-tidy [hunk 1]
```diff
@@ -2,5 +2,5 @@
 # FIXME: all performance-* reports
 # FIXME: all cert-* reports
 # FIXME: all bugprone-* reports
-Checks: -*,bugprone-*,-bugprone-unhandled-self-assignment,-bugprone-parent-virtual-call,-bugprone-narrowing-conversions,-bugprone-exception-escape,-bugprone-string-literal-with-embedded-nul,cppcoreguidelines-slicing,mpi-*,readability-non-const-parameter,performance-*,modernize-*,-modernize-use-trailing-return-type,-modernize-use-bool-literals,-modernize-avoid-c-arrays,-modernize-use-auto,-modernize-return-braced-init-list,-bugprone-easily-swappable-parameters
+Checks: -*,bugprone-*,-bugprone-unhandled-self-assignment,-bugprone-parent-virtual-call,-bugprone-narrowing-conversions,-bugprone-exception-escape,-bugprone-string-literal-with-embedded-nul,cppcoreguidelines-slicing,mpi-*,readability-non-const-parameter,performance-*,-performance-avoid-endl,modernize-*,-modernize-use-trailing-return-type,-modernize-use-bool-literals,-modernize-avoid-c-arrays,-modernize-use-auto,-modernize-return-braced-init-list,-bugprone-easily-swappable-parameters,-bugprone-chained-comparison
```
Comment (ax3l) [inline .clang-tidy:5]: Please rebase :)

Comment (franzpoeschel) [reply .clang-tidy:5]: done

File: test/SerialIOTest.cpp [hunk 2]
```diff
@@ -5856,21 +5864,21 @@ void variableBasedSeries(std::string const &file)
             // this tests changing extents across iterations
             // ADIOS2 does not support changing the dimensionality
             // (older versions used to somewhat support it, but not really)
-            auto E_y = iteration.meshes["E"]["y"];
+            auto B_y = iteration.meshes["B"]["y"];
```
Comment (ax3l) [inline test/SerialIOTest.cpp:5867]: Interesting example. I was thinking primarily about particle records, but yes fields also could be used to shapes of some components depend on others...

**Review by ax3l — APPROVED** [review]
